### PR TITLE
400 add math ops and refactored fixed point config

### PR DESCRIPTION
**Pull request by LeoBuron**
This feature adds basic support for future use of quantized fixed point gradients. It includes a refactor of the existing a new  fixed point config which is refactored because the old one was doing to much. Following, I changed the naming a bit and have now a fixed point configuration with datafields only. The quantization function is now moved to the quantize_to_fixed_point file using the fixed point configuration.

### CODE REVIEW

File: elasticai/creator/nn/quantized_grads/fixed_point/_quantize_to_fixed_point.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+from elasticai.creator.nn.quantized_grads.fixed_point import FixedPointConfigV2
+
+
+def quantize(number: torch.Tensor, fxp_conf: FixedPointConfigV2) -> torch.Tensor:
+    return _round(clamp(number, fxp_conf), fxp_conf)
+
+
+def _round(number: torch.Tensor, fxp_conf: FixedPointConfigV2) -> torch.Tensor:
+    if fxp_conf.stochastic_rounding:
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/_quantize_to_fixed_point.py:11]: That if statement will not be found during tracing, therefore we might prefer to use two different functions. Following this thought, we'd end up with two different models, one using fxp rounding while the other doesn't. Not sure if that'd be ok for you, but i think it's the cleaner approach. That way you'd not even have to add the stochastic rounding flag to fxp_config and the fxp_config could just be `(total_Bits, frac_bits)`.

Also, from experience flag arguments like this often make it hard to know which rounding method is actually used, especially if the `fxp_conf` object is mutable.

For now i think it'd even be fine to just add a comment, that you'd have to be aware of this when tracing. Or we flag entire modules as not being fit for tracing. Maybe @mokouMonday has an opinion here (i guess you're one of the people for which that matters).

Rest looks good ;).

File: elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import torch
+
+from ._quantize_to_fixed_point import quantize
+from ._two_complement_fixed_point_config import FixedPointConfigV2
+
+
+class RoundToFixedPoint(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx: Any, *args: Any, **kwargs: Any) -> torch.Tensor:
+        if len(args) != 3:
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py:12]: you won't catch the case where people supply more parameters as kwargs, but I don't even think you need to handle that here

File: elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            raise TypeError(
+                "apply() takes exactly two arguments "
+                "(x: torch.Tensor, config: FixedPointConfig)"
+            )
+        x: torch.Tensor = args[0]
+        fxp_config: FixedPointConfigV2 | None = args[1]
+        ctx.grad_fxp_config: FixedPointConfigV2 | None = args[2]
+        if fxp_config is None:
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py:20]: This instead might be an actual problem. The RoundToFixedPoint function not rounding to fixed point but instead acting as an identity seems like a Liskov violation to me. I'd never expect this function to not somehow round a value. We could actually raise an error in case no fxp config is given.

File: elasticai/creator/nn/quantized_grads/fixed_point/_two_complement_fixed_point_config.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from dataclasses import dataclass
+
+import torch
+
+
+@dataclass
+class FixedPointConfigV2:
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/_two_complement_fixed_point_config.py:7]: we could add a

```python
class FixedPointConfig:
  __slots__ = ("frac_bits", "total_bits", "stochastic_rounding")
  
  def __init__(self, total_bits: int, frac_bits: int, stochastic_rounding: bool):
    #...
```

This prevents adding more fields to an object of this type after creation and makes attribute access way faster. I.e.,

```python
fxp = FixedPointConfig(4, 2, False)
fxp.something_else = 4 # runtime error
```
Finally, since this lives in the quantized_grads fixed_point package rn, you could just call this `FixedPointConfig` as well, no problem. Another options is: we place it in the same package as the old `FixedPointConfig` and call it `FixedPointConfigV2`.

I think both of these would make it easier to replace the old config by your newer one.

**Review by glencoe — CHANGES_REQUESTED** [review]
The `if` branches (at least the one in the rounding function) I'd like to see addressed. Everything else I mention is just comments i guess.

Oh... almost forgot: Not sure there are tests for the old fxp config (i fear not). Do you want to add a few for the rounding behaviour? I mean a lot of the rest here is hard to test and not really test worthy, i guess.

File: elasticai/creator/nn/quantized_grads/fixed_point/_quantize_to_fixed_point.py (hunk 1 above)
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/_quantize_to_fixed_point.py:11]: I made the FixedPointConfigV2 immutable. I changed the functions, so tracing will allow to see if deterministic or stochastic rounding is used.

File: elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py (hunk 2 above)
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py:12]: I added the catch

File: elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py (hunk 3 above)
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py:20]: While I agree with your take that this function should from the naming scheme never do this, I would really like to have a chance to inherit no FxpConfig to be able to not round something. But I have no idea how to do this easily. Do you have an idea?

File: elasticai/creator/nn/quantized_grads/fixed_point/_two_complement_fixed_point_config.py (hunk 4 above)
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/_two_complement_fixed_point_config.py:7]: I fixed it with making the class immutable. This way I also catch your worries from the other comment.

File: elasticai/creator/nn/quantized_grads/fixed_point/_quantize_to_fixed_point.py (hunk 1 above)
Comment (glencoe) [reply elasticai/creator/nn/quantized_grads/fixed_point/_quantize_to_fixed_point.py:7]: Making it immutable is certainly an improvement here, but i have the impression that you might misunderstand how tracing works. What i wanted to say above is that tracing will not see that there is an if statement in python code. I think immutability solves the issue of the traced version not behaving like the code it was traced from though.

File: elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    quantize
+
+def test_round_determinstic_to_fixed_point():
+    conf = FixedPointConfigV2(total_bits=8,
+                           frac_bits=2,
+                           stochastic_rounding=False)
+    x = torch.range(-2, 2, step=0.1, dtype=torch.float32)
+
+    actual = round_to_fixed_point(x, conf.frac_bits)
+    expected = torch.Tensor([-2.0000, -2.0000, -1.7500, -1.7500, -1.5000, -1.5000, -1.5000, -1.2500,
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py:14]: What's the reason for listing four digits after decimal point?

File: elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    x = torch.range(-3, 3, step=1, dtype=torch.float32)
+
+    actual = clamp(x, conf)
+    expected = torch.Tensor([-2.0000, -2.0000, -1.0000,  0.0000,  1.0000,  1.7500,  1.7500])
+    assert(torch.equal(actual, expected))
+
+def test_quantize_determinstic_to_fixed_point():
+    conf = FixedPointConfigV2(total_bits=4,
+                              frac_bits=2,
+                              stochastic_rounding=False)
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py:35]: The *deterministic* part of unit tests sucks a bit here, right? 
I'm currently eyeballing with https://hypothesis.readthedocs.io/en/latest/
it leverages pytest and automatically generates input data for your algorithm, it will then remember for which input data your tests failed and be able to replay these. I think for all the numeric stuff we need to do in the creator and with ML it could be really powerful.

Especially recommend this article for an introduction into their stateful testing: https://hypothesis.works/articles/how-not-to-die-hard-with-hypothesis/

File: elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py (hunk 3 above)
Comment (glencoe) [reply elasticai/creator/nn/quantized_grads/fixed_point/_round_to_fixed_point_autograd.py:26]: Not sure, what you mean with "inherit" in this context. But here's how Julian or me would probably do it.
Instead of injecting a boolean parameter, we inject the function itself. Somewhere you will call `RoundToFixedPoint.apply(x, None)` in a layer. Let's say that layer is called `Conv`. So `Conv` looks like this

```python
class Conv(torch.nn.Conv):
  def __init__(self, fxp_config: FxpConfig | None):
    super().__init__()
    self._fxp_config = fxp_config

  def forward(self, x):
    weights = RoundToFixedPoint.apply(self.weights, self._fxp_config)
    return torch.nn.functional.conv(x, weights)
```

Instead you write

```python
class Conv(torch.nn.Conv):
  def __init__(self, quantize: Callable[[Tensor], Tensor]):
    super().__init__()
    self._quantize = quantize

  def forward(self, x):
    weights = quantize(self.weights, self._fxp_config)
    return torch.nn.functional.conv(x, weights)

def round_to_fixed_point(x: Tensor, fxp_config: FxpConfig) -> Tensor:
  return RoundToFixedPoint.apply(x, fxp_config)

from functools import partial

my_config = FxpConfig(total_bits=4, frac_bits=3)
c = Conv(quantize=partial(
        round_to_fixed_point, 
        fxp_config=my_conf))
c_without_quant = Conv(quantize=lambda x: x)
```

**Review by glencoe — COMMENTED** [review]
A few comments and a hopefully useful explanation on dependency inversion via dependency injection.

File: elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py (hunk 5 above)
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py:14]: No reason. Changed it back.

File: elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py (hunk 6 above)
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py:35]: Why does it suck? 
The deterministic part is that I only test deterministic rounding which implements Round half to even. 

The tool you mention sounds cool. I guess it would make sense to have look at it with a small demo. 
But should be a different PR

Comment (glencoe) [reply elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py:35]: > But should be a different PR

Of course it should.

I always find testing stochastic processes difficult, you want to make sure that your sample size is big enough to cover a meaningful number of cases, on the other hand you want deterministic behaviour, so you can rely on your tests.

File: elasticai/creator/nn/quantized_grads/fixed_point/_two_complement_fixed_point_config.py [hunk 7]
```diff
@@ -1,13 +1,16 @@
...
+from elasticai.creator.nn.quantized_grads.quantization_config import QuantizationConfig
+
 
 @dataclass(frozen=True)
-class FixedPointConfigV2:
+class FixedPointConfigV2(QuantizationConfig):
     total_bits: int
     frac_bits: int
-    stochastic_rounding: bool
+    stochastic_rounding: Optional[bool] = False
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/fixed_point/_two_complement_fixed_point_config.py:13]: With optional you're saying that this may be None. Also note: this syntax for optionals is deprecated (and now i understand why :D), we should use `bool | None` instead.

File: elasticai/creator/nn/quantized_grads/quantize.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+from jedi.inference.gradual.typing import Protocol
+
+from elasticai.creator.nn.quantized_grads.quantization_config import QuantizationConfig
+
+
+class Quantize(Protocol):
+    config: QuantizationConfig
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/quantize.py:7]: Having that field here seems odd. Can't we just call the quantize function from outside so that it quantizes? What i mean is, is there a use case, where another caller has to access that config? If so, then why are we not defining anything else about it? I mean there's nothing a caller could do with `QuantizationConfig` except for passing it around, because the corresponding dataclass is empty.

**Review by glencoe — COMMENTED** [review]
lgtm looks good to me

**Review by glencoe — APPROVED** [review]
